src/network.rs

```diff
@@ -87,140 +95,80 @@ impl Network {
         Distribution::new(self.sections.keys().map(|prefix| prefix.len() as u64))
     }
 
-    fn generate_random_messages(&mut self) {
-        let mut adds = 0;
-        let mut drops = 0;
 
-        for section in self.sections.values_mut() {
-            if random::gen() {
-                add_random_node(&self.params, section);
-                adds += 1;
-
-                if drop_random_node(&self.params, section) {
-                    drops += 1;
-                }
-            } else {
-                if drop_random_node(&self.params, section) {
-                    drops += 1;
-                }
-
-                add_random_node(&self.params, section);
-                adds += 1;
-            }
-        }
-
-        info!(
-            "Random Adds: {} Drops: {}",
-            log::important(adds),
-            log::important(drops)
-        );
-    }
-
-    fn handle_messages(&mut self) -> TickStats {
-        let mut responses = Vec::new();
+    fn handle_actions(&mut self, actions: &mut Vec<Action>) -> TickStats {
         let mut stats = TickStats::new();
 
-        loop {
-            for section in self.sections.values_mut() {
-                responses.extend(section.handle_requests(&self.params));
-            }
-
-            if responses.is_empty() {
-                break;
-            }
-
-            stats += self.handle_responses(&mut responses)
-        }
+        for action in actions.drain(..) {
+            match action {
+                Action::Reject(_) => {
+                    stats.rejections += 1;
+                }
+                Action::Merge(target) => {
+                    let sources: Vec<_> = self.sections
+                        .keys()
+                        .filter(|prefix| prefix.is_descendant(&target))
+                        .cloned()
+                        .collect();
+
+                    if sources.is_empty() {
+                        continue;
```

Comment:
better also have a log and comment here to be consistent with the case in split below

Reply:
:+1: